c++/src/kj/async-io-test.c++

```diff
@@ -3574,5 +3574,203 @@ KJ_TEST("pump file to socket") {
+
+  writePromise.wait(io.waitScope);
+  auto amount = readPromise2.wait(io.waitScope);
+  KJ_ASSERT(amount == 5);
+  KJ_ASSERT(memcmp(buffer, "hello", 5) == 0);
+
+  abortThread.detach();
+}
+#endif  // !_WIN32
+
+// ---------------------------------------------------------------------------------------------
+// accept() with aborted connection - dual-stack IPv4/IPv6 listener
+// ---------------------------------------------------------------------------------------------
+// Test accept() behavior with aborted cross-protocol connections on dual-stack listeners.
+// Creates an aborted IPv4 connection to an IPv6 listener, followed by a valid IPv6 connection.
+//
+// Expected behavior:
+// - Darwin/macOS: When IPv4 connects to IPv6 listener and gets aborted, accept() returns
+//   a socket with addrlen=0. KJ's accept loop detects this Darwin quirk and discards the
+//   socket automatically, so first accept() returns the valid connection.
+// - Linux/other Unix: accept() returns the aborted connection, which fails on first
+//   read (throws exception or returns 0 bytes). Test then calls accept() again for the
+//   valid connection.
+//
+// This test specifically exercises the Darwin addrlen==0 bug workaround in KJ's accept loop.
+
+#if !_WIN32
+KJ_TEST("accept() with aborted connection - dual-stack IPv4/IPv6") {
+  if (!systemSupportsAddress("::")) {
+    KJ_LOG(WARNING, "system does not support ipv6; skipping test");
+    return;
+  }
+  char buffer[16];
+  // -- async-io boilerplate ---------------------------------------------------
+  auto io = kj::setupAsyncIo();
+  auto& net = io.provider->getNetwork();
+
+  auto listenerAddr = net.parseAddress("::", 0).wait(io.waitScope);
+  auto listener = listenerAddr->listen();
+  uint16_t port = listener->getPort();
+
+  // Create IPv4 connection that will be aborted (sends RST packet)
+  kj::Thread abortThread([&] {
+    int s = ::socket(AF_INET, SOCK_STREAM, 0);
+    KJ_ASSERT(s >= 0);
+
+    // Configure socket to send RST on close instead of FIN
+    struct linger lg = {1, 0};
+    KJ_SYSCALL(setsockopt(s, SOL_SOCKET, SO_LINGER, &lg, sizeof(lg)));
+
+    sockaddr_in a {};
+    a.sin_family = AF_INET;
+    a.sin_port = htons(port);
+    a.sin_addr.s_addr = htonl(0x7f000001);
+    KJ_SYSCALL(::connect(s, reinterpret_cast<sockaddr*>(&a), sizeof(a)));
+
+    // Close immediately to send RST packet
+    KJ_SYSCALL(::close(s));
+  });
+
+  // Allow aborted connection to reach accept queue first
+  io.provider->getTimer().afterDelay(2 * kj::MILLISECONDS)
+                         .wait(io.waitScope);
+
+  // Create valid IPv6 connection
+  auto clientP = net.parseAddress("::1", port)
+      .then([](Own<NetworkAddress> addr) { return addr->connect(); });
+  // Accept connection - behavior differs by platform
+  auto serverCon = listener->accept().wait(io.waitScope);
+  KJ_ASSERT(serverCon);
+
+#if __APPLE__
+  // On Apple platforms: IPv4->IPv6 aborted connections return addrlen=0,
```

Comment:
This is effectively testing for the XNU bug. If they fix the bug this test may start failing.

I think if you just delete the __APPLE__ side of the branch here, and use the other side for all platforms, it should work? Because that code is OK if the platform filters out the aborted connection. But it would still fail without your fix, right?